// react-amphtml/dist/setup.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));
var __chunk_1 = require('./chunk-874ba39f.js');
require('./chunk-00861ebd.js');
var amphtml = require('./amphtml.js');

const AmpScripts = class {
  constructor(htmlFormat = __chunk_1.AMP) {
    this.scripts = new Map([[htmlFormat, React__default.createElement(amphtml.Script, {
      key: htmlFormat,
      specName: htmlFormat
    })]]);
  }

  addExtension({
    extension,
    version
  }) {
    this.scripts.set(extension, React__default.createElement(amphtml.Script, {
      key: extension,
      specName: extension,
      version: version
    }));
  }

  getScriptElements() {
    return [...this.scripts.values()];
  }

};

const AmpScriptsManager = class extends React.Component {
  getChildContext() {
    const {
      ampScripts
    } = this.props;
    return {
      [__chunk_1.CONTEXT_KEY]: ampScripts
    };
  }

  render() {
    const {
      children
    } = this.props;
    return React__default.Children.only(children);
  }

};
AmpScriptsManager.childContextTypes = {
  [__chunk_1.CONTEXT_KEY]: PropTypes.instanceOf(AmpScripts).isRequired
};
AmpScriptsManager.propTypes = {
  children: PropTypes.node.isRequired,
  ampScripts: PropTypes.instanceOf(AmpScripts).isRequired
};

/* eslint-disable react/no-danger */

const key = suffix => `react-amphtml-header-boilerplate-${suffix}`;

var headerBoilerplate = (href => [React__default.createElement("meta", {
  key: key('meta-charset'),
  charSet: "utf-8"
}), React__default.createElement("meta", {
  key: key('meta-viewport'),
  name: "viewport",
  content: "width=device-width,minimum-scale=1,initial-scale=1"
}), React__default.createElement("link", {
  key: key('canonical-link'),
  rel: "canonical",
  href: href
}), React__default.createElement("style", {
  key: key('style'),
  "amp-boilerplate": "",
  dangerouslySetInnerHTML: {
    __html: `
          body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
        `
  }
}), React__default.createElement("noscript", {
  key: key('noscript')
}, React__default.createElement("style", {
  "amp-boilerplate": "",
  dangerouslySetInnerHTML: {
    __html: `
            body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}
          `
  }
}))]);

exports.AmpScripts = AmpScripts;
exports.AmpScriptsManager = AmpScriptsManager;
exports.headerBoilerplate = headerBoilerplate;
